tools/notification-configuration/notification-creator/NotificationConfigurator.cs

```diff
@@ -149,12 +159,97 @@ public NotificationConfigurator(AzureDevOpsService service, ILogger<Notification
                 if (persistChanges)
                 {
                     result = await service.UpdateTeamForProjectAsync(pipeline.Project.Id.ToString(), result);
+                    if (purpose == TeamPurpose.ParentNotificationTeam)
+                    {
+                        await EnsureScheduledBuildFailSubscriptionExists(pipeline, result, true);
+                    }
                 }
             }
 
+            if (purpose == TeamPurpose.SynchronizedNotificationTeam)
+            {
+                await SyncTeamWithCodeOwnerFile(pipeline, result, gitHubToAADConverter, persistChanges);
+            }
             return result;
         }
 
+        private async Task SyncTeamWithCodeOwnerFile(BuildDefinition pipeline, WebApiTeam team, GitHubToAADConverter gitHubToAADConverter, bool persistChanges)
+        {
+            using (logger.BeginScope("Team Name = {0}", team.Name))
+            {
+                if (pipeline.Process.Type != PipelineYamlProcessType)
+                {
+                    return;
+                }
+
+                // Get contents of CODEOWNERS
+                logger.LogInformation("Fetching CODEOWNERS file");
+                var managementUrl = new Uri(pipeline.Repository.Properties["manageUrl"]);
+                var codeOwnerEntries = await gitHubService.GetCodeownersFile(managementUrl);
+
+                if (codeOwnerEntries == default)
+                {
+                    logger.LogInformation("CODEOWNERS file not found, skipping sync");
+                    return;
+                }
+
+                var process = pipeline.Process as YamlProcess;
+
+                logger.LogInformation("Searching CODEOWNERS for matching path for {0}", process.YamlFilename);
+                var codeOwnerEntry = CodeOwnersFile.FindOwnersForClosestMatch(codeOwnerEntries, process.YamlFilename);
+                codeOwnerEntry.FilterOutNonUserAliases();
+
+                logger.LogInformation("Matching Contacts Path = {0}, NumContacts = {1}", process.YamlFilename, codeOwnerEntry.Owners.Count);
+
+                // Get set of team members in the CODEOWNERS file
+                var codeownerPrincipals = codeOwnerEntry.Owners
+                    .Select(contact => {
```

Comment:
We chatted offline about this but summarizing here. I'm a little concerned about the cache update side effect in this LINQ statement, given that it's lazy evaluated and in general LINQ operations are not recommended for side effects. I'm not sure if that would cause issues here, but I think it might make sense to investigate when/whether we actually get cache hits (and whether they are deterministic or not) when executing this async below.

Reply:
Yes we shouldn't do this in a linq expression. Just turn this into a foreach loop and then add to the list as you go.

Reply:
Use foreach loop as it lose the select syntax benefits.